File: Forgex_Build/license-tools/FxLicenseCore/Services/MachineFingerprintService.cs
using System.Net;
using System.Net.NetworkInformation;
using System.Security.Cryptography;
using System.Text;

namespace FxLicenseCore.Services;

/// <summary>
/// 机器指纹服务。
/// </summary>
/// <remarks>
/// 与后端 MachineFingerprintService 保持一致，
/// 用于在请求授权客户端生成稳定机器码。
/// </remarks>
public sealed class MachineFingerprintService
{
    /// <summary>
    /// 解析机器码。
    /// </summary>
    /// <param name="overrideMachineCode">覆盖机器码，可为空。</param>
    /// <returns>SHA-256 机器码。</returns>
    public string ResolveMachineCode(string? overrideMachineCode = null)
    {
        if (!string.IsNullOrWhiteSpace(overrideMachineCode))
        {
            return Normalize(overrideMachineCode);
        }

        var parts = new List<string>();
        var osName = Environment.OSVersion.Platform.ToString().ToLowerInvariant();

        if (osName.Contains("win", StringComparison.Ordinal))
        {
            AddIfHasText(parts, Environment.GetEnvironmentVariable("COMPUTERNAME"));
            AddIfHasText(parts, Environment.GetEnvironmentVariable("PROCESSOR_IDENTIFIER"));
            AddIfHasText(parts, Environment.GetEnvironmentVariable("SystemDrive"));
        }
        else
        {
            AddIfHasText(parts, ReadFileQuietly("/etc/machine-id"));
            AddIfHasText(parts, ReadFileQuietly("/sys/class/dmi/id/product_uuid"));
            AddIfHasText(parts, Environment.GetEnvironmentVariable("HOSTNAME"));
        }

        AddIfHasText(parts, ResolveHostname());
        parts.AddRange(ResolveMacAddresses());

        if (parts.Count == 0)
        {
            parts.Add("FORGEX-UNKNOWN-HOST");
        }

        var joined = string.Join("|", parts).ToUpperInvariant();
        using var sha256 = SHA256.Create();
        var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(joined));
        return Convert.ToHexString(hash);
    }

    /// <summary>
    /// 获取当前主机名。
    /// </summary>
    /// <returns>主机名。</returns>
    public string ResolveHostname()
    {
        try
        {
            return Dns.GetHostName();
        }
        catch
        {
            return "unknown-host";
        }
    }

    private static void AddIfHasText(List<string> parts, string? value)
    {
        if (!string.IsNullOrWhiteSpace(value))
        {
            parts.Add(Normalize(value));
        }
    }

    private static string Normalize(string value)
    {
        return value.Trim().Replace(" ", string.Empty, StringComparison.Ordinal).ToUpperInvariant();
    }

    private static string? ReadFileQuietly(string path)
    {
        try
        {
            return File.Exists(path) ? File.ReadAllText(path).Trim() : null;
        }
        catch
        {
            return null;
        }
    }

    private static IEnumerable<string> ResolveMacAddresses()
    {
        try
        {
            return NetworkInterface.GetAllNetworkInterfaces()
                .Where(item => item.OperationalStatus == OperationalStatus.Up)
                .Where(item => item.NetworkInterfaceType != NetworkInterfaceType.Loopback)
                .Where(item => item.NetworkInterfaceType != NetworkInterfaceType.Tunnel)
                .Select(item => item.GetPhysicalAddress()?.ToString())
                .Where(item => !string.IsNullOrWhiteSpace(item))
                .Cast<string>()
                .ToList();
        }
        catch
        {
            return [];
        }
    }
}

File: Forgex_MOM/Forgex_Backend/Forgex_Common_Infra/src/main/java/com/forgex/common/license/MachineFingerprintService.java
package com.forgex.common.license;

import cn.hutool.crypto.digest.DigestUtil;
import org.springframework.util.StringUtils;

import java.net.InetAddress;
import java.net.NetworkInterface;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Enumeration;
import java.util.List;
import java.util.Locale;

/**
 * 机器指纹生成服务
 * <p>
 * 负责生成当前宿主机的稳定机器码，优先使用安装器注入值，其次回退到宿主机特征拼装。
 * </p>
 *
 * @author Forgex Team
 * @version 1.0.0
 */
public class MachineFingerprintService {

    private final LicenseProperties licenseProperties;

    public MachineFingerprintService(LicenseProperties licenseProperties) {
        this.licenseProperties = licenseProperties;
    }

    /**
     * 获取当前机器码。
     *
     * @return SHA-256 机器码
     */
    public String resolveMachineCode() {
        if (StringUtils.hasText(licenseProperties.getOverrideMachineCode())) {
            return normalize(licenseProperties.getOverrideMachineCode());
        }

        List<String> parts = new ArrayList<>();
        String osName = System.getProperty("os.name", "").toLowerCase(Locale.ROOT);

        if (osName.contains("win")) {
            addIfHasText(parts, System.getenv("COMPUTERNAME"));
            addIfHasText(parts, System.getenv("PROCESSOR_IDENTIFIER"));
            addIfHasText(parts, System.getenv("SystemDrive"));
        } else {
            addIfHasText(parts, readFileQuietly("/etc/machine-id"));
            addIfHasText(parts, readFileQuietly("/sys/class/dmi/id/product_uuid"));
            addIfHasText(parts, System.getenv("HOSTNAME"));
        }

        addIfHasText(parts, resolveHostName());
        parts.addAll(resolveMacAddresses());

        if (parts.isEmpty()) {
            parts.add("FORGEX-UNKNOWN-HOST");
        }

        String joined = String.join("|", parts).toUpperCase(Locale.ROOT);
        return DigestUtil.sha256Hex(joined.getBytes(StandardCharsets.UTF_8)).toUpperCase(Locale.ROOT);
    }

    private void addIfHasText(List<String> parts, String value) {
        if (StringUtils.hasText(value)) {
            parts.add(normalize(value));
        }
    }

    private String normalize(String value) {
        return value == null ? "" : value.trim().replace(" ", "").toUpperCase(Locale.ROOT);
    }

    private String readFileQuietly(String filePath) {
        try {
            Path path = Path.of(filePath);
            if (!Files.exists(path)) {
                return null;
            }
            return Files.readString(path).trim();
        } catch (Exception ignored) {
            return null;
        }
    }

    private String resolveHostName() {
        try {
            return InetAddress.getLocalHost().getHostName();
        } catch (Exception ignored) {
            return null;
        }
    }

    private List<String> resolveMacAddresses() {
        try {
            List<String> macs = new ArrayList<>();
            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
            if (interfaces == null) {
                return macs;
            }
            for (NetworkInterface networkInterface : Collections.list(interfaces)) {
                if (!networkInterface.isUp() || networkInterface.isLoopback() || networkInterface.isVirtual()) {
                    continue;
                }
                byte[] hardwareAddress = networkInterface.getHardwareAddress();
                if (hardwareAddress == null || hardwareAddress.length == 0) {
                    continue;
                }
                StringBuilder sb = new StringBuilder();
                for (byte b : hardwareAddress) {
                    sb.append(String.format("%02X", b));
                }
                if (sb.length() > 0) {
                    macs.add(sb.toString());
                }
            }
            return macs;
        } catch (Exception ignored) {
            return Collections.emptyList();
        }
    }
}
